DOCSP-37469: Add transactions page (#199)
mongodb/docs-csharp: source/fundamentals/transactions.txt

**transactions**. :manual:`Transactions </core/transactions/>` allow
you to run a series of operations that do not change any data until the
transaction is committed. If any operation in the transaction returns an
error, the driver cancels the transaction and discards all data changes
before they ever become visible.

In MongoDB, transactions run within logical **sessions**. A
:manual:`session </reference/server-sessions/>` is a grouping of related
read or write operations that you intend to run sequentially. Sessions
enable :manual:`causal consistency
</core/read-isolation-consistency-recency/#causal-consistency>` for a
group of operations or allow you to execute operations in an
:website:`ACID transaction </basics/acid-transactions>`. MongoDB
guarantees that the data involved in your transaction operations remains
consistent, even if the operations encounter unexpected errors.

When using the {+driver-short+}, you can create a new session from a
``MongoClient`` instance as an ``IClientSession`` type. We recommend that you reuse
your client for multiple sessions and transactions instead of
instantiating a new client each time.

.. warning::

>    Use an ``IClientSession`` only with the ``MongoClient`` (or associated

Reviewer: Let's start using

   ``MongoDatabase`` or ``MongoCollection``) that created it. Using an
   ``IClientSession`` with a different ``MongoClient`` results in operation
   errors.

Methods
-------

Create an ``IClientSession`` by using either the synchronous ``StartSession()`` or
the asynchronous ``StartSessionAsync()`` method on your ``MongoClient`` instance. 
You can then modify the session state by using the method set
provided by the ``IClientSession`` interface. Select from the following
:guilabel:`Synchronous Methods` and :guilabel:`Asynchronous Methods`
tabs to learn about the methods to manage your transaction:

.. tabs::

    .. tab:: Synchronous Methods
        :tabid: synchronous-methods

        .. list-table::
            :widths: 40 60
            :header-rows: 1

            * - Method
              - Description

            * - ``StartTransaction()``

>               - | Starts a new transaction, configured with the given options, on

Reviewer: Let's mention that `TransactionOptions` are optional here.

                  this session. Returns an error if there is already
                  a transaction in progress for the session. To learn more about
                  this method, see the :manual:`startTransaction() page
                  </reference/method/Session.startTransaction/>`  in the Server manual.
                |
                | **Parameter**: ``TransactionOptions``

>             * - ``AbortTransaction()``

Reviewer: We don't return an error but throw an exception instead. This is applicable for all methods.

error, the driver cancels the transaction and discards all data changes
before they ever become visible.

In MongoDB, transactions run within logical **sessions**. A
:manual:`session </reference/server-sessions/>` is a grouping of related
read or write operations that you intend to run sequentially. Sessions
enable :manual:`causal consistency
</core/read-isolation-consistency-recency/#causal-consistency>` for a
group of operations or allow you to execute operations in an
:website:`ACID transaction </basics/acid-transactions>`. MongoDB
guarantees that the data involved in your transaction operations remains
consistent, even if the operations encounter unexpected errors.

When using the {+driver-short+}, you can create a new session from a
``MongoClient`` instance as an ``IClientSession`` type. We recommend that you reuse
your client for multiple sessions and transactions instead of
instantiating a new client each time.

.. warning::

   Use an ``IClientSession`` only with the ``MongoClient`` (or associated
   ``MongoDatabase`` or ``MongoCollection``) that created it. Using an
   ``IClientSession`` with a different ``MongoClient`` results in operation
   errors.

Methods
-------

> Create an ``IClientSession`` by using either the ``StartSession()`` or
> ``StartSessionAsync()`` methods on your ``MongoClient`` instance, depending on your
> application's usecase. You can then modify the session state by using the method set

Reviewer 2: ```suggestion
Create an ``IClientSession`` by using either the synchronous ``StartSession()`` or
the asynchronous ``StartSessionAsync()`` method on your ``MongoClient`` instance. You can then modify the session state by using the method set
```

provided by the ``IClientSession`` interface. Select from the following
:guilabel:`Synchronous Methods` and :guilabel:`Asynchronous Methods`

> tabs to learn about the methods to manage your transaction that correspond to each
> framework:

Reviewer 2: S: maybe we can remove the end of the sentence as it is obvious
```suggestion
tabs to learn about the methods to manage your transaction:
```

                | **Return Type**: ``Task``

            * - ``WithTransactionAsync()``
              - | Starts a transaction on this session and runs the given callback. To
                  learn more about this method, see the :manual:`withTransaction() page
                  </reference/method/Session.withTransaction/>` in the Server manual.

                .. warning:: Handling Exceptions

                   When catching exceptions within the callback function used by
                   ``WithTransactionAsync()``, you  **must** rethrow the exception
                   before exiting the try-catch block. Failing to do so can result in an
                   infinite loop. For further details on how to handle exceptions in this
                   case, see :manual:`Transactions </core/transactions/>` in the Server
                   manual and select :guilabel:`C#` from the language dropdown to view 
                   the example.

                |
                | **Parameters**: ``Func <IClientSessionHandle, CancellationToken, Task<TResult>>``, ``TransactionOptions``, ``CancellationToken``
                | **Return Type**: ``Task <TResult>``

Example
-------

> This example shows how you can create a session, create a
> transaction, and commit an insert operation into multiple collections through
> the following steps:

Reviewer 2: ```suggestion
This example shows how you can create a session, create a
transaction, and insert documents into multiple collections within the transaction
through the following steps:
```

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about any of the types or methods discussed in this
guide, see the following API Documentation:

- `IClientSession <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IClientSession.html>`__
- `MongoClient <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.MongoClient.html>`__
- `StartTransaction() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IClientSession.StartTransaction.html>`__
- `AbortTransaction() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IClientSession.AbortTransaction.html>`__ / `AbortTransactionAsync() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IClientSession.AbortTransactionAsync.html>`__
- `CommitTransaction() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IClientSession.CommitTransaction.html>`__ / `CommitTransactionAsync() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IClientSession.CommitTransactionAsync.html>`__
- `WithTransaction() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IClientSession.WithTransaction.html>`__ / `WithTransactionAsync() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IClientSession.WithTransactionAsync.html>`__

> - `TransactionOptions() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.TransactionOptions.html>`__

Reviewer 2: ```suggestion
- `TransactionOptions <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.TransactionOptions.html>`__
```
